simpeg/potential_fields/magnetics/simulation.py

```diff
@@ -1730,497 +1800,387 @@ def getB0(self):
+        return np.concatenate(total_data_list, axis=0)
 
-        fields = {}
-        if "bx" in components or "tmi" in components:
-            fields["bx"] = self.Qfx
-        if "by" in components or "tmi" in components:
-            fields["by"] = self.Qfy
-        if "bz" in components or "tmi" in components:
-            fields["bz"] = self.Qfz
+    @utils.count
+    def _projectFieldsDeriv(self, Bs):
+        components = self.survey.components
 
         if "tmi" in components:
-            bx = fields["bx"]
-            by = fields["by"]
-            bz = fields["bz"]
-            # Generate unit vector
             B0 = self.survey.source_field.b0
             Bot = np.sqrt(B0[0] ** 2 + B0[1] ** 2 + B0[2] ** 2)
-            box = B0[0] / Bot
-            boy = B0[1] / Bot
-            boz = B0[2] / Bot
-            fields["tmi"] = bx * box + by * boy + bz * boz
-
-        return sp.vstack([fields[comp] for comp in components])
-
-    def _get_components(self):
-        """
-        Get components of all receivers in the survey.
 
-        This function assumes that all receivers in the survey have the same
-        components in the same order.
+            bx = self._Qfx * Bs
+            by = self._Qfy * Bs
+            bz = self._Qfz * Bs
 
-        Returns
-        -------
-        components : list of str
-            List of components shared by all receivers in the survey.
-
-        Raises
-        ------
-        ValueError
-            If the survey doesn't have any receiver, or if any receiver has
-            a different set of components than the rest.
-        """
-        # Validate survey first to ensure that the receivers have all the same
-        # components.
-        self._validate_survey(self.survey)
-        components = self.survey.source_field.receiver_list[0].components
-        return components
-
-    def _validate_survey(self, survey):
-        """
-        Validate a survey for the magnetic differential 3D simulation.
-
-        Parameters
-        ----------
-        survey : Survey
-            Survey object that will get validated.
-
-        Raises
-        ------
-        ValueError
-            If the survey doesn't have any receiver, or if any receiver has
-            a different set of components than the rest.
-        """
-        receivers = survey.source_field.receiver_list
-        if not receivers:
-            msg = "Found invalid survey without receivers."
-            raise ValueError(msg)
-        components = receivers[0].components
-        if not all(components == rx.components for rx in receivers):
-            msg = (
-                "Found invalid survey with receivers that have mixed components. "
-                f"Surveys for the {type(self).__name__} class must contain receivers "
-                "with the same components."
+            dpred = (
+                np.sqrt((bx + B0[0]) ** 2 + (by + B0[1]) ** 2 + (bz + B0[2]) ** 2) - Bot
```

Comment:
Code duplication with _projectFields. Suggest creating a private method to extract components giving fields.
Something along those lines
```
def _components(self, fields) -> dict[str, np.ndarray]:
    """
    Extract components from magnetic fields.
    """
    B0 = self.survey.source_field.b0
    comps = {
    "bx": self._Qfx * fields,
    "by": self._Qfy * fields,
    "bz": self._Qfz * fields,
    "b_total": np.linalg.norm(B0)
     }
     
     comps["tmi"] =  np.sqrt(
                (comps["bx"] + B0[0]) ** 2 + (comps["by"] + B0[1]) ** 2 + (comps["bz"] + B0[2]) ** 2
            ) - comps["b_total"]
     return comps
```
so you can just call it in both methods